### PR TITLE
chore(web): restores previous test-setup patterns for Web unit tests 🎡

### PR DESCRIPTION
**Pull request by jahorton**
Reverts the "poor man's modules" approach used by #6597 and #6622, allowing the pure-headless unit tests for the following packages to work via Node modules again.

Currently done for the following packages:
- `keyboard-processor`
- `input-processor`

This is primarily accomplished via the new `tsconfig.bundled.json` configuration files for the affected projects.  These rely near-completely on the 'standard' configuration's settings... just with all relevant references `prepend`ed, as for testing... these need to be bundled as leaf products.  To module-include the build-product, it needs to be in a fully-compact "leaf product" form.

TypeScript project references don't yet have a `#pragma once` analogue ([reference](https://www.typescriptlang.org/docs/handbook/project-references.html#prepend-with-outfile)), but some of these packages have (internal) dependences that see multiple inclusions.  (`keyboard-processor` reuses some of the same components that the LMLayer uses.)

Furthermore, when the `lm-worker` part is needed, that requires some extra handling... so I just copied the "TEMP STUB" hacky bit found elsewhere in the 🎡 chain.  (Compilation targeting the DOM doesn't look kindly on Worker-oriented globals.)

There are a few other Web subpackages in need of similar treatment.

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by jahorton** [comment]
Looks like there may be a bit of build agent weirdness on the 🎡 chain - despite touching nothing that would affect iOS, some of its unit tests are failing here, but not on its base branch.

**Comment by mcdurdin** [comment]
> Looks like there may be a bit of build agent weirdness on the 🎡 chain - despite touching nothing that would affect iOS, some of its unit tests are failing here, but not on its base branch.

```
Failed at the @keymanapp/lm-worker@ clean script.
```

Ah, kmsample1 build is not running npm ci at the start of the build process...

File: common/core/web/input-processor/test.sh
Comment (mcdurdin) [inline common/core/web/input-processor/test.sh:59]:
```suggestion
./build.sh
```

`./build.sh` to keep it consistent with our build processes? Also, this should be optional -- we shouldn't force a rebuild for a test run... But perhaps I'll resolve this particular aspect with the build.sh refactoring coming soon (which moves test.sh into `build.sh test`, where a build would happen with `build.sh build test`.

File: common/core/web/keyboard-processor/tests/cases/basic-init.js
Comment (mcdurdin) [inline common/core/web/keyboard-processor/tests/cases/basic-init.js:13]: Can you turn on an option in your editor to trim whitespace on end of lines? That avoids this kind of whitespace back-and-forth :grin:

**Review by mcdurdin — COMMENTED** [review]
This generally looks good. I'm not a big fan of forcing a build before a test -- it should be up to the developer -- but am comfortable with leaving that pattern until we get to the build.sh refactor later.

I am wondering if the bundled versions of tsconfig should be in a separate folder -- or at least whether the `outFile` folder/`outDir` should be differentiated, to avoid potential conflicts and inconsistent builds?

I'll hold off approval until we've got the current chain merged down to a single PR in #6525, then we'll revisit this along with #6625 to resolve merge conflicts, and merge them into that PR before we do the full regression test run.

File: common/core/web/input-processor/test.sh
Comment (jahorton) [reply common/core/web/input-processor/test.sh:59]: > we shouldn't force a rebuild for a test run

But we should force a build when it's never been built at all.  Which is what we'd expect to see on a CI build, no?  Builds of `input-processor` (for Web) are separate of those for the lexical model compiler (for Developer).

Might be nice if we had some clean and simple "build if unbuilt" command.  Do we have any existing patterns for that?

Comment (mcdurdin) [reply common/core/web/input-processor/test.sh:59]: > But we should force a build when it's never been built at all. Which is what we'd expect to see on a CI build, no?

A CI build should be explicit in what it is doing: `build.sh configure build test`. So we don't need to worry about that particular scenario. We should avoid implicit behaviours in CI as far as we can (this includes dependency install, builds, getting correct tool versions on PATH, etc).

But I am planning on implementing some of the checks in the build.sh refactor work I will be doing soon. There's a bit of reinvent-the-wheel in all this because that's what build orchestrators are supposed to be good at, but I have not found any build orchestrators that cover all the different build environments we work in, and a simple file existence function in build-utils.sh will probably do enough of what we need.

Long term, we may be able to use another build orchestrator, but I think that consistent build.sh bash-based scripts for now will get us a very long way forward.

Comment (jahorton) [reply common/core/web/input-processor/test.sh:59]: > A CI build should be explicit in what it is doing: build.sh configure build test. 

Okay then - how would you propose that the CI build for Web _explicitly_ specifies that all test dependencies are built in Web's `build.sh`, as that's the script that will trigger all these other test scripts?

Comment (jahorton) [reply common/core/web/input-processor/test.sh:59]: Let me get a little explicit here.

At present, all Web-related tests are triggered via a _single_ `test.sh` call:  `web/unit_tests/test.sh`.  That fans out in a dependency-based tree:

- `web/unit_tests/test.sh`
    - `common/core/web/input_processor/test.sh`
        - `common/predictive-text/unit_tests/test.sh`
          - Also requires `lm-worker`.
        - `common/core/web/keyboard_processor/test.sh`
        - Also requires `developer/js/build.sh`, `lm-worker`
- Also requires `common/core/web/tools/recorder/src/build.sh`, `lm-worker`

I get the feeling that to achieve what you say you want, we'll need to break the link between the individual `test.sh` scripts and have a 'master' test script that defines & runs the full tree?  We'd definitely want to keep said 'master' script in repo, though, as we're slowly adding more web packages, and we can't go leaving out components by accident.

Comment (mcdurdin) [reply common/core/web/input-processor/test.sh:59]: Ideally, in the end, we will have pretty much the same outcome as you currently have with `web/unit_tests/test.sh`, but we will instead call `web/build.sh [configure build] test`. (`configure` and `build` may or may not be necessary, depending on where in the CI chain it is called.) The way in which we define the hierarchy may change, but the general principle of each platform having a single script call to run all tests relevant to that platform will remain.

As to how we get there, I don't have a complete answer yet. This documentation of the very complex web of web interdependencies helps.

Some random thoughts from a tired brain:
* All scripts will be in repo. CI will only call the scripts.
* A unit test in /common having a dependency on /developer is a bit problematic (potential for circular dependencies etc) -- this is part of a larger issue where we have cross-product (particularly developer) dependencies.
* Ideally, I want each component to have its own unit tests, which can be run without reference to any other component (otherwise it is not a unit test -- it's an end-to-end test).
* Multi-module or global or end-to-end tests: I am still trying to figure out where we should put these, particularly ones that interact across multiple products.
* I think we do give each web module a `build.sh` script. This script will do `build`, preferably `test`, perhaps `configure`. It will also eventually define the module's dependencies (which we may be able to extract from tsconfig.json for web projects).
* While lerna needs to be replaced, the replacement I've chosen (npm workspaces + tsc projects) sadly duplicates dependency information in `package.json` and `tsconfig.json` 🤮. So I am seriously keen to avoid a third lot of duplication! (Weird idea from fevered imagination: what if `package.json` and/or `tsconfig.json` were generated in the `configure` step?!?)

**Comment by jahorton** [comment]
Oh good, we didn't lose the important comments after the rebase.  I was worried that may happen.

> This generally looks good. I'm not a big fan of forcing a build before a test -- it should be up to the developer -- but am comfortable with leaving that pattern until we get to the build.sh refactor later.

By the way, I took a considerable amount of time looking for _some_ sort of `tsc` flag that could override the reference `prepend` setting to prepend everything.  Unfortunately... it doesn't seem that such a compiler flag is available; it'd have been the easiest way to avoid the dual-tsconfig approach used here.

> I am wondering if the bundled versions of tsconfig should be in a separate folder -- or at least whether the outFile folder/outDir should be differentiated, to avoid potential conflicts and inconsistent builds?

I did wonder if I should add a nested folder within the base output directory for the test-bundled version.  I'm not opposed to that approach.

**Comment by jahorton** [comment]
Hmm.  For the plain iOS target build...

```
[Step 3/5]     resultString = "Your session has expired. Please log in.";
[14:13:40]
[Step 3/5]     userLocale = "en_US";
[14:13:40]
[Step 3/5]     userString = "Your session has expired. Please log in.";
[14:13:40]
[Step 3/5] }, NSLocalizedDescription=Your session has expired. Please log in.}
```

... yay.

For the Samples one... something's squirrel-y there.  It does try to compile, but it's getting odd messages that I can't repro locally.

File: common/web/input-processor/src/tsconfig.bundled.json
Comment (mcdurdin) [inline common/web/input-processor/src/tsconfig.bundled.json:17]:
```suggestion
    { "path": "../../es6-shim", "prepend": true }
```

**Review by mcdurdin — APPROVED** [review]
I think this is ok -- we need to consider the dependencies on developer in common tests in the future but for now we can go with it.

**Comment by jahorton** [comment]
Huh... looks like the BrowserStack tests aren't being run correctly.

Fortunately, that helped me notice something else in the logs - I almost disconnected the `keyboard-processor` unit tests!  Glad that got caught.

**Comment by jahorton** [comment]
And I think the iOS one is related to the new FV certs?

**Comment by jahorton** [comment]
Yep, somehow I'd missed the certs commit.

**Comment by jahorton** [comment]
Just tested the BrowserStack testing configuration from my machine; it works when hosting locally.  So, I'm not currently able to locally repro whatever issue is occurring.

**Pull request merged**
